NuRadioReco/modules/channelSinewaveSubtraction.py

```
@@ -0,0 +1,271 @@
+    """
+
+    dt = 1 / sampling_rate # in ns
+    t = np.arange(0, len(wf) * dt, dt) # in ns
+
+    # zero meaning, just in case
+    wf = wf - np.mean(wf)
+
+    def sinusoid(t, amplitude, noise_frequency, phase):
+        return amplitude * np.sin(2 * np.pi * noise_frequency * t + phase + np.pi/2)
+
+    spec = abs(fft.time2freq(wf, sampling_rate))
+    freqs = fft.freqs(len(wf), sampling_rate)
+
+    # find noise frequencies:
+    rms = np.sqrt(np.mean(np.abs(spec) ** 2))
+    peak_idxs = np.where(np.abs(spec) > peak_prominance * rms)[0]
+
+    noise_freqs = []
+    corrected_waveform = wf.copy()
+    # find mean CW freq bean
+    if len(peak_idxs) > 0:
+
+        # Initialize a list to hold groups of neighboring peak indices
+        group = [peak_idxs[0]]
+
+        # Loop through the remaining peak indices to group neighboring peaks
+        for i in range(1, len(peak_idxs)):
+            if peak_idxs[i] - peak_idxs[i - 1] == 1:  # If the peak is neighboring the previous one
+                group.append(peak_idxs[i])
+            else:
+                # Calculate the mean frequency for the current group of neighbors
+                noise_freqs.append(np.mean(freqs[group]))
+                # Start a new group with the current peak
+                group = [peak_idxs[i]]
+
+        # Don't forget to append the last group
+        if group:
+            noise_freqs.append(np.mean(freqs[group]))
+
+        # Convert the list to a NumPy array (optional, if you prefer an array)
+        noise_freqs = np.array(noise_freqs)
+
+        if saved_noise_freqs is not None:
+
+            saved_noise_freqs.append(noise_freqs)
+
+        for noise_freq in noise_freqs:
+
+            ampl_guess = guess_amplitude(wf, noise_freq, sampling_rate)
+            initial_guess = [ampl_guess, noise_freq, 0.01]
+
+            # Fit the sinusoidal model to the waveform
+            try:
+                params, _ = curve_fit(sinusoid, t, wf, p0=initial_guess)
```

check if fit converged? Or does curve_fit throw exceptions on lack of convergence? (that would be weird, but I suppose it's possible... )

---

curve_fit should raise a RuntimeError if minimization fails. I added checks to see if the fit parameters or the covariance matrix are invalid.

Do you think I should remove the exception and just let the script crash if the fit fails? I thought it be better to just keep going and log the info.

---

Ah, ok, yes that would be surprising behavior if the script crashes because the minimization fails (though, from looking at the code, only some failures result in exceptions as far as I can tell)). I agree it's better to let it fail and log, though you may want to add some sort of flag to th event as well (if you don't already, I don't remember).